### DISCUSSION TITLE
Trying to get openblas installed, with fortran to install scipy on windows arm64

### DISCUSSION THREAD

**Discussion by Molier** [OpenMathLib/OpenBLAS#4914] [Q&A] [answered]
Trying to install scipy on windows arm64, surface pro11, Snapdragon(R) X 12-core X1E80100 @ 3.40 GHz, 3417 Mhz, 12 Core(s), 12 Logical Processor(s).

i have installed gcc, mingw and all the needed cpp build tools from visual studio installer as well. Also llvM-18.1.8-woa64.exe .

Close to a complete dev setup. but now Im not finding an arm64 windows build of FORTRAN or GFORTRAN anywhere. I want to run as much as possible natively. But a bit stuck atm.
```scoop list
Installed apps:

Name           Version            Source   Updated             Info
----           -------            ------   -------             ----
7zip           24.08              main     2024-10-01 16:01:14
cmake          3.30.4             main     2024-10-03 10:07:14
dark           3.14               main     2024-10-01 15:10:56 64bit
gcc            13.2.0             main     2024-10-03 10:00:27 64bit
innounp        0.50               main     2024-10-01 16:01:17 64bit
mingw          14.2.0-rt_v12-rev0 main     2024-10-03 10:08:10 64bit
nilesoft-shell 1.9.15             extras   2024-10-01 16:01:20 64bit
pipx           1.7.1              main     2024-10-01 15:10:44 64bit
python         3.12.6             main     2024-10-01 15:11:22
python311      3.11.9             versions 2024-10-01 15:15:56
rust           1.81.0             main     2024-10-03 09:29:44
```
found some steps highlighted [here in this stack overflow ](https://stackoverflow.com/questions/75762998/how-to-install-scipy-and-numpy-on-windows-on-arm64)as well: 

> ... To build windows+ARM native packages of numpy and scipy:
> 
> Not only will you need a C/C++ compiler, you will also need a Fortran compiler. I am not sure if gfortran is available for WinARM (that's what we typically use to build scipy wheels). ifort is another compiler to look into. scipy wheels are made with gcc/g++/gfortran.
> You will need to download + build a BLAS library from source. The scipy/numpy projects typically use OpenBLAS.
> You will need to build + install packages for all the build time dependencies. i.e. Cython, pybind11, (pythran), meson, meson-python, ninja. Some of those may have dependencies of their own. A couple of those have OS/arch independent wheels, but not all.
> Once you've built all of those you'll build numpy, then scipy.....

**Comment by martin-frbg** [comment]
I can't help with your quest for a WoA gfortran as I do not have such a system, but it is possible to build OpenBLAS without a Fortran compiler - the only drawback is that you'll get a somewhat older version of LAPACK. (The current codebase of the Reference-LAPACK project has added some functions in modern Fortran dialects that cannot as easily be translated to C).

**Reply by Molier** [reply]
mmmh interesting, but as I have LLVM installed, that includes [a version of flang ](https://github.com/llvm/llvm-project/tree/main/flang)correct? so why can I not use that to then build openblas/scipy
for reference here is the current error im having:
```
- Installing scipy (1.14.1): Failed

  ChefBuildError

  Backend subprocess exited when trying to invoke build_wheel
  
  + meson setup C:\Users\user\AppData\Local\Temp\tmp5orumm7b\scipy-1.14.1 C:\Users\user\AppData\Local\Temp\tmp5orumm7b\scipy-1.14.1\.mesonpy-n57c3w0a -Dbuildtype=release -Db_ndebug=if-release -Db_vscrt=md --native-file=C:\Users\user\AppData\Local\Temp\tmp5orumm7b\scipy-1.14.1\.mesonpy-n57c3w0a\meson-python-native-file.ini
  The Meson build system
  Version: 1.5.2
  Source dir: C:\Users\user\AppData\Local\Temp\tmp5orumm7b\scipy-1.14.1
  Build dir: C:\Users\user\AppData\Local\Temp\tmp5orumm7b\scipy-1.14.1\.mesonpy-n57c3w0a
  Build type: native build
  Project name: scipy
  Project version: 1.14.1
  Activating VS 17.11.4
  C compiler for the host machine: cl (msvc 19.41.34120 "Microsoft (R) C/C++ Optimizing Compiler Version 19.41.34120 for ARM64")
  C linker for the host machine: link link 14.41.34120.0
  C++ compiler for the host machine: cl (msvc 19.41.34120 "Microsoft (R) C/C++ Optimizing Compiler Version 19.41.34120 for ARM64")
  C++ linker for the host machine: link link 14.41.34120.0
  Cython compiler for the host machine: cython (cython 3.0.11)
  Host machine cpu family: aarch64
  Host machine cpu: aarch64
  Program python found: YES (C:\Users\user\AppData\Local\Temp\tmps89m9jjv\.venv\Scripts\python.exe)
  Run-time dependency python found: YES 3.11
  Program cython found: YES (C:\Users\user\AppData\Local\Temp\tmps89m9jjv\.venv\Scripts\cython.EXE)
  Compiler for C supports arguments -Wno-unused-but-set-variable: NO 
  Compiler for C supports arguments -Wno-unused-function: NO 
  Compiler for C supports arguments -Wno-conversion: NO 
  Compiler for C supports arguments -Wno-misleading-indentation: NO 
  Library m found: NO
  
  ..\meson.build:84:0: ERROR: Unknown compiler(s): [['ifort'], ['gfortran'], ['flang'], ['pgfortran'], ['g95']]
  The following exception(s) were encountered:
  Running `ifort --version` gave "[WinError 2] The system cannot find the file specified"
  Running `ifort -V` gave "[WinError 2] The system cannot find the file specified"
  Running `gfortran --version` gave "[WinError 2] The system cannot find the file specified"
  Running `gfortran -V` gave "[WinError 2] The system cannot find the file specified"
  Running `flang --version` gave "[WinError 2] The system cannot find the file specified"
  Running `flang -V` gave "[WinError 2] The system cannot find the file specified"
  Running `pgfortran --version` gave "[WinError 2] The system cannot find the file specified"
  Running `pgfortran -V` gave "[WinError 2] The system cannot find the file specified"
  Running `g95 --version` gave "[WinError 2] The system cannot find the file specified"
  Running `g95 -V` gave "[WinError 2] The system cannot find the file specified"
  
  A full log can be found at C:\Users\user\AppData\Local\Temp\tmp5orumm7b\scipy-1.14.1\.mesonpy-n57c3w0a\meson-logs\meson-log.txt
```

I need to get 1 of the options working I suppose

**Comment by martin-frbg** [comment]
I'm not sure if the precompiled LLVM package provided by the LLVM project contains any version of flang or flang-new - the one for Windows x86 definitely does not. Do you have a flang.exe or flang-new.exe ?
Unfortunately you'll have to ask the Scipy people about any problems with their meson-based build system. OpenBLAS itself does not use this so far (though there have been talks about adding support for it)

**Reply by Molier** [reply]
the precompiled package contains flang-new.exe yes https://github.com/llvm/llvm-project/releases/tag/llvmorg-18.1.8 thw WoA version

**Comment by martin-frbg** [comment]
Interesting - in that case it could simply be that the meson script is only looking for it as "flang" and not the current, likely temporary name of flang-new. But as I said I know next to nothing about their current meson setup - there is cooperation between our projects but not much direct involvement in the innards of each other's code.
Guess you could try copying flang-new.exe to its old name to get around the detection problem, but this might not be sufficient if it does not handle behavioral differences between the old and new flang

**Comment by martin-frbg** [comment]
I had a quick look at the scipy build files but did not see anything that would do compiler name detection there - this appears to be a core functionality of meson. Indeed the meson repository reveals that initial support for flang-new only went into the very latest release, 1.5.2 from two weeks ago (though the actual source commit was in July, right after the 1.5.1 release). This appears to be the one that you are using, but it could be that this new feature is not working correctly. Maybe try setting `FC=flang-new.exe` in the environment, or open an issue ticket at either https://github.com/mesonbuild/meson/issues or https://github.com/scipy/scipy/issues

**Reply by Molier** [reply]
https://pastebin.com/zamjnLxx error log here
I have installed llvm ( thus including flang) using the latest [19.1.0](https://snapshots.linaro.org/llvm-toolchain/19/rc1/), I also built it myself from source to test it, but that didnt really seem to work. I think now I have a decent version of flang detected. but now its asking for openblas if I read the error log correctly. I installed it from here: [armdevelopers](https://developer.arm.com/Tools%20and%20Software/Arm%20Performance%20Libraries) but I dont think its being correctly detected or something. I cant seem to find a good software stack that is working for me on my arm64 :( but I saw that it must be able to work and build, since another user succeeded [here](https://github.com/scipy/scipy/issues/20801).

**Comment by martin-frbg** [comment, marked as answer]
That is not even OpenBLAS but ARM's libarmpl. _Please_ go ask over in scipy/scipy how to resolve your installation problems with their software

**Reply by Molier** [reply]
thanks for your help, will create an issue there. Will also try to build openblas myself then, [these docs ](https://github.com/OpenMathLib/OpenBLAS/wiki/How-to-build-OpenBLAS-for-Windows-on-ARM64)look like good starting point.
